core/modules/wpublish/QueriesAndChunks.cc

```diff
@@ -0,0 +1,486 @@
+/// The sub-maps contain information about how long tasks take to complete on that table
+/// on that chunk. These are then used to see what the percentage total of the time it took
+/// for tasks on each chunk.
+/// The table names are based on the slowest scan table in each task.
+QueriesAndChunks::ScanTableSumsMap QueriesAndChunks::_calcScanTableSums() {
+    // Copy a vector of all the chunks in the map;
+    std::vector<ChunkStatistics::Ptr> chks;
+    {
+        std::lock_guard<std::mutex> g(_chunkMtx);
+        for (auto const& ele : _chunkStats) {
+            auto const& chk = ele.second;
+            chks.push_back(chk);
+        }
+    }
+
+    QueriesAndChunks::ScanTableSumsMap scanTblSums;
+    for (auto const& chunkStats : chks) {
+        auto chunkId = chunkStats->_chunkId;
+        std::lock_guard<std::mutex> lock(chunkStats->_tStatsMtx);
+        for (auto const& ele : chunkStats->_tableStats) {
+            auto const& tblName = ele.first;
+            if (tblName != "") {
+                auto &sTSums = scanTblSums[tblName];
+                sTSums.totalTime  += ele.second->getAvgCompletionTime();
+                ChunkTimePercent &ctp = sTSums.chunkPercentages[chunkId];
+                ctp.shardTime = ele.second->getAvgCompletionTime();
+            }
+        }
+    }
+
+    // Calculate percentage totals.
+    for (auto &eleTbl : scanTblSums) {
+        auto &scanTbl = eleTbl.second;
+        auto totalTime = scanTbl.totalTime;
+        for (auto &eleChunk : scanTbl.chunkPercentages) {
+            auto &tPercent = eleChunk.second;
+            tPercent.percent = tPercent.shardTime / totalTime;
+        }
+    }
+    return scanTblSums;
+}
+
+
+/// Remove the running 'task' from a scheduler and possibly move all Tasks that belong to its user query
+/// to the snail scheduler. 'task' continues to run in its thread, but the scheduler is told 'task' is
+/// finished, which allows the scheduler to move on to another Task.
+/// If too many Tasks from the same user query are booted, all remaining tasks are moved to the snail
+/// scheduler in an attempt to keep a single user query from jamming up a scheduler.
+void QueriesAndChunks::_bootTask(QueryStatistics::Ptr const& uq, wbase::Task::Ptr const& task,
+                                     wsched::SchedulerBase::Ptr const& sched) {
+    LOGS(_log, LOG_LVL_INFO, task->getIdStr() << " taking too long, booting from " << sched->getName());
+    sched->removeTask(task);
+    uq->_tasksBooted += 1;
+    if (uq->_tasksBooted > _maxTasksBooted) {
+        LOGS(_log, LOG_LVL_INFO, task->getIdStr() << " entire UserQuery booting from " << sched->getName());
+        removeQueryFrom(uq->_queryId, sched);
+    }
+}
+
+
+/// Add a Task to the user query statistics.
+void QueryStatistics::addTask(wbase::Task::Ptr const& task) {
+    std::lock_guard<std::mutex> guard(_qStatsMtx);
+    std::pair<int, wbase::Task::Ptr> ent(task->getJobId(), task);
+    _taskMap.insert(ent);
```

Comment:
This one-liner looks more readable to me (but it may be only me :cat:):

    _taskMap.insert(std::make_pair(task->getJobId(), task));